### PR TITLE
Adapted MemRegion::getDescriptiveName to handle ElementRegions

### LINKED ISSUE #84463: [analyzer] MemRegion::getDescriptiveName breaks for Index not of type nonloc::ConcreteInt
I am currently working with the Clang Static Analyzer to analyze small C code examples. Thereby I have come across a behaviour of MemRegion::getDescriptiveName that I do not fully understand. I am working with the llvm-project 17.x-release version. 
The simple checker below should be sufficient for demonstration purposes. Basically, the MemRegion and its name for a given SVal should be extracted within the checkLocation method.

```Cpp
namespace {
class TestChecker : public clang::ento::Checker<clang::ento::check::Location> {

public:
  void checkLocation(const clang::ento::SVal &Loc, bool IsLoad,
                     const clang::Stmt *S,
                     clang::ento::CheckerContext &CC) const {
    clang::ento::ProgramStateRef PSR = CC.getState();
    const clang::ento::MemRegion *MemReg = Loc.getAsRegion();

    if (MemReg) {
      std::string MemRegName = MemReg->getDescriptiveName();
      llvm::outs() << "MemRegion: Name = " << MemRegName << "\n";
    }
  }
};
} // namespace

void clang::ento::registerTestChecker(clang::ento::CheckerManager &CM) {
  CM.registerChecker<TestChecker>();
}

bool clang::ento::shouldRegisterTestChecker(
    const clang::ento::CheckerManager &CM) {
  return true;
}
```

This is the C code example to be analysed.

```C
unsigned int index1 = 1;
extern unsigned int index2;
extern int array[10];

int main() {
  array[index1];
  array[index2];
  return 0;
}
```

The extracted MemRegions and their names for the first statement match my expectations.:
```
MemRegion: Name = 'index1'
ElementRegion: ER = Element{array,1 S64b,int}
Index: ConcreteInt = 1
Next iteration: R = array
MemRegion: Name = 'array[1]'
```

Here is the relevant part of MemRegion::getDescriptiveName with a few additional outputs:
```Cpp
std::string MemRegion::getDescriptiveName(bool UseQuotes) const {
  std::string VariableName;
  std::string ArrayIndices;
  const MemRegion *R = this;
  SmallString<50> buf;
  llvm::raw_svector_ostream os(buf);

  // Obtain array indices to add them to the variable name.
  const ElementRegion *ER = nullptr;
  while ((ER = R->getAs<ElementRegion>())) {
    // Index is a ConcreteInt.
    llvm::outs() << "ElementRegion: ER = " << ER->getString() << "\n";
    if (auto CI = ER->getIndex().getAs<nonloc::ConcreteInt>()) {
      llvm::SmallString<2> Idx;
      CI->getValue().toString(Idx);
      ArrayIndices = (llvm::Twine("[") + Idx.str() + "]" + ArrayIndices).str();
      llvm::outs() << "Index: ConcreteInt = " << Idx.str() << "\n";
    }
    // If not a ConcreteInt, try to obtain the variable
    // name by calling 'getDescriptiveName' recursively.
    else {
      std::string Idx = ER->getDescriptiveName(false);
      if (!Idx.empty()) {
        ArrayIndices = (llvm::Twine("[") + Idx + "]" + ArrayIndices).str();
        llvm::outs() << "Index: NOT ConcreteInt = " << Idx << "\n";
      }
    }
    R = ER->getSuperRegion();
    if (R)
      llvm::outs() << "Next iteration: R = " << R->getString() << "\n";
  }
...
```

The second statement, however, causes the method to call itself recursively with the same instance (else branch inside while statement). This finally leads to a segmentation fault.
```
MemRegion: Name = 'index2'
ElementRegion: ER = Element{array,reg_$0<unsigned int index2>,int}
ElementRegion: ER = Element{array,reg_$0<unsigned int index2>,int}
ElementRegion: ER = Element{array,reg_$0<unsigned int index2>,int}
...
#254 0x000055f02a6171ed clang::ento::MemRegion::getDescriptiveName[abi:cxx11](bool) const (bin/clang+0x57321ed)
#255 0x000055f02a6171ed clang::ento::MemRegion::getDescriptiveName[abi:cxx11](bool) const (bin/clang+0x57321ed)
Segmentation fault (core dumped)
```

The result that would correspond to my expectation could be achieved by the following additional lines (else if):

```Cpp
std::string MemRegion::getDescriptiveName(bool UseQuotes) const {
  std::string VariableName;
  std::string ArrayIndices;
  const MemRegion *R = this;
  SmallString<50> buf;
  llvm::raw_svector_ostream os(buf);

  // Obtain array indices to add them to the variable name.
  const ElementRegion *ER = nullptr;
  while ((ER = R->getAs<ElementRegion>())) {
    // Index is a ConcreteInt.
    llvm::outs() << "ElementRegion: ER = " << ER->getString() << "\n";
    if (auto CI = ER->getIndex().getAs<nonloc::ConcreteInt>()) {
      llvm::SmallString<2> Idx;
      CI->getValue().toString(Idx);
      ArrayIndices = (llvm::Twine("[") + Idx.str() + "]" + ArrayIndices).str();
      llvm::outs() << "Index: ConcreteInt = " << Idx.str() << "\n";
    }
    // Added: Try to get index as SymbolVal -> SymbolRef -> OriginRegion -> Name
    else if (auto SI = ER->getIndex().getAs<nonloc::SymbolVal>()) {
      if (auto SR = SI->getAsSymbol()) {
        if (auto OR = SR->getOriginRegion()) {
          std::string Idx = OR->getDescriptiveName(false);
          ArrayIndices = (llvm::Twine("[") + Idx + "]" + ArrayIndices).str();
          llvm::outs() << "Index: SymbolVal = " << Idx << "\n";
        }
      }
    }
    // If not a ConcreteInt, try to obtain the variable
    // name by calling 'getDescriptiveName' recursively.
    else {
      std::string Idx = ER->getDescriptiveName(false);
      if (!Idx.empty()) {
        ArrayIndices = (llvm::Twine("[") + Idx + "]" + ArrayIndices).str();
        llvm::outs() << "Index: NOT ConcreteInt = " << Idx << "\n";
      }
    }
    R = ER->getSuperRegion();
    if (R)
      llvm::outs() << "Next iteration: R = " << R->getString() << "\n";
  }
  ```

This would achieve the following output and the recursion of MemRegion::getDescriptiveName would also be solved for this case:
 ```
 MemRegion: Name = 'index2'
ElementRegion: ER = Element{array,reg_$0<unsigned int index2>,int}
Index: SymbolVal = index2
Next iteration: R = array
MemRegion: Name = 'array[index2]'
```

Am I missing something or is the described behaviour intended?

### PR DESCRIPTION
**Pull request by T-Gruber**
Fixes https://github.com/llvm/llvm-project/issues/84463

Changes:
- Adapted MemRegion::getDescriptiveName
- Added unittest to check name for a given clang::ento::ElementRegion
- Some format changes due to clang-format

### CODE REVIEW

**Comment by steakhal** [comment]
I hope you don't mind that I reworked the testing a bit, to be similar to the rest of the tests.
I also reverted unnecessary clang-formatting where it was not relevant.
I refined the `auto` usage, but other than that I left your code untouched.

As I did everything I thought, I'll invite others to review the PR.

**Comment by T-Gruber** [comment]
Hi @steakhal, thank you very much for your additional changes!

**Comment by T-Gruber** [comment]
Hello @steakhal, I have just looked through the changes again. What is the advantage of using checkPreCall instead of checkLocation? I would very much appreciate some background information. Thanks for your help!

**Comment by steakhal** [comment]
> Hello @steakhal, I have just looked through the changes again. What is the advantage of using checkPreCall instead of checkLocation? I would very much appreciate some background information. Thanks for your help!

Oh yes, I should have explained.
So, `checkLocation` can be triggered for all loads and store operations. Each time in the AST you have a `LValueToRValue` cast for reads, and others for write operations. This means that you can't really define a test where you have clear control of what region you wanted to dump.

Using `PreCall`, you can have your own API, that does what you think, exactly when it encounters that call.
This is similar to what we do in the `ExprInspection` checker, where we have analyzer debug intrinsics defined. Such as `clang_analyzer_dump(T)`. Search for it in the tests, and you will see what I'm talking about.

There is one more benefit of doing this way: There is already a communication channel for the stringified descriptive name: diagnostics.
One can easily capture and compare the diagnostics, like in the rest of the tests.

I was happy to help as I don't expect new contributors to be really involved with our specific testing.
To me what moved the needle was: well described expectation, well described and reproducible actual outcome. You even provided a fix, so I figured I'll help out with the rest.

Ask me if there is anything I left out.

**Comment by T-Gruber** [comment]
Thanks again @steakhal! This is helpful and interesting background information. I really appreciate your help!

**Review by NagyDonat — CHANGES_REQUESTED** [review]
I strongly suspect that there are some situations where this commit would behave incorrectly (assert on legal code or print incorrect output). See inline comment for explanation and a suggested solution (that I did not test).

File: clang/lib/StaticAnalyzer/Core/MemRegion.cpp [hunk 1]
```diff
@@ -720,14 +720,20 @@ std::string MemRegion::getDescriptiveName(bool UseQuotes) const {
...
+          std::string Idx = OR->getDescriptiveName(false);
+          ArrayIndices = (llvm::Twine("[") + Idx + "]" + ArrayIndices).str();
+        }
       }
     }
+    // Index is neither a ConcreteInt nor SymbolVal, give up and return.
+    else {
+      assert(false && "we should have a descriptive name");
+      return "";
+    }
```
[the remark this answers is not in the document]
Comment (steakhal) [reply clang/lib/StaticAnalyzer/Core/MemRegion.cpp:736]: Let me elaborate what I think.
To set the scene, I'd like to point out that there was an unconditional infinite recursion in the previous code; and we in practice never seen crashes related to this so I think it's fair to say that currently we don't have any checkers exercising this path. This suggests to me that we could basically do anything here, that is better than a certain crash (aka. improve the previous handling).

So, at this point I'm okay to accept any code that at least sometimes work - given that we have a test for exercising that path.

And btw, if any of the sub-conditions fail, we hit the `else` branch and we assert in debug builds to get notified, and silently ignore the case (and may produce incomplete messages that people can still notice and report) in release builds.
Consequently, this shouldn't affect users in a blocking way.

**Comment by steakhal** [comment]
> In addition to the above-mentioned issues there is also a conceptual problem with using `getOriginRegion()` to describe a symbol: it names the region where the symbol _originated_ (if it originated as the unknown initial value of a symbol), which is not necessarily the region where the symbol is stored _now_.
> 
> For example consider the (somewhat contrived) code
> 
> ```
> int matrix[10][10];
> int func(int x, int y) {
>   int tmp = x;
>   x = y;
>   y = tmp;
>   return matrix[x][-999]; 
> }
> ```
> 
> Here the checker alpha.security.ArrayBoundV2 (which uses `getDescriptiveName`) would report that
> 
> ```
> Access of 'matrix[y]' at negative byte offset
> ```
> 
> because the symbolic value of the index _originated from_ the parameter `y` (i.e. the symbol is identified as "this is the initial value of `y`).

I agree, but I don't see how this relates to this patch.
I think the original author of this code probably wanted to something what we propose here.
We can argue if that's a good thing or not, but that feels like a different topic to me.

File: clang/lib/StaticAnalyzer/Core/MemRegion.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (T-Gruber) [reply clang/lib/StaticAnalyzer/Core/MemRegion.cpp]: Hi @NagyDonat, thank you for the hint. I improved the implementation based on your suggestion and also added 3 more test cases to the unittest.

**Comment by T-Gruber** [comment]
> In addition to the above-mentioned issues there is also a conceptual problem with using `getOriginRegion()` to describe a symbol: it names the region where the symbol _originated_ (if it originated as the unknown initial value of a symbol), which is not necessarily the region where the symbol is stored _now_.
> 
> For example consider the (somewhat contrived) code
> 
> ```
> int matrix[10][10];
> int func(int x, int y) {
>   int tmp = x;
>   x = y;
>   y = tmp;
>   return matrix[x][-999]; 
> }
> ```
> 
> Here the checker alpha.security.ArrayBoundV2 (which uses `getDescriptiveName`) would report that
> 
> ```
> Access of 'matrix[y]' at negative byte offset
> ```
> 
> because the symbolic value of the index _originated from_ the parameter `y` (i.e. the symbol is identified as "this is the initial value of `y`).

Yes, thats right. But the previous implementation ended up in an endless recursion if the index is not of type ConcreteInt. Therefore, the proposed version is a significant improvement (at least for my use cases).

**Review by NagyDonat — APPROVED** [review]
LGTM, thanks for adding the testcases!

I agree that the "sometimes reports the wrong variable" situation is a very significant improvement over the old "always infinite loop" situation; but this "convert a symbol to a variable name" step was completely missing from the old code, so I think it's relevant to mention that this commit implements it in a way that's not entirely correct.

I think it would be good to highlight this limitation with a FIXME comment (and perhaps a testcase), but I don't want to significantly delay the resolution of the infinite loop bug.

**Comment by steakhal** [comment]
@T-Gruber Could you please add that test case with a FIXME?

**Comment by T-Gruber** [comment]
Hi @steakhal, just to be sure. We are talking about a test case that covers that the index is "recognised" as a wrong variable (the matrix example above)?

**Comment by steakhal** [comment]
> Hi @steakhal, just to be sure. We are talking about a test case that covers that the index is "recognised" as a wrong variable (the matrix example above)?

Exactly!

**Comment by NagyDonat** [comment]
Since suggesting that "matrix" testcase I realized that it won't work, because `ArrayBoundV2` doesn't perform bounds checking for the lower-dimensional sub-arrays of a multidimensional array (it only checks that the accessed memory location is within the full array). For example in `int matrix[10][10];` it reports that `matrix[1][200]` overflows the bounds of "matrix", but doesn't report `matrix[1][50]`. I tried to eliminate this limitation, but that patch had to be reverted because the meaning of `ElementRegion` is ambiguous.

Due to this limitation, it's a bit difficult to ensure that `ArrayBoundV2` calls `getDescriptiveName()` on an `ElementRegion`, but I think adding a `FieldRegion` layer should work:
```
struct {
  int numbers[10];
} table[100];
int func(int x, int y) {
  int tmp = x;
  x = y;
  y = tmp;
  return table[x].numbers[-1]; 
}
```
Here I expect that the checker would (incorrectly) report something like `Access of 'table[y].numbers' at negative byte offset`.

**Comment by T-Gruber** [comment]
Alright, I added a test case which leads to an incorrect index due to the use of getOriginRegion. I also added a FIXME comment to document that this case is not covered currently.

**Review by NagyDonat — APPROVED** [review]
LGTM.

Using this unit test framework is significantly cleaner than my suggestion that tries to work around the limitations of ArrayBoundV2.

**Pull request merged**